### ISSUE TITLE
Decryption of encrypted payloads stopped working

### ISSUE THREAD

**Issue by harbra** [owntracks/recorder#348]
I had a runnting ot-recorder installation for some time now. In April apparently decryption of encrypted payloads stopped working for me (and I didn't realize). The ot-recorder log shows `no decryption key for XXX-YYY in owntracks/XXX/YYY` with encrypted payloads, however I'm not aware I changed anything in my setup (apart from the MQTT server domain). Also `ocat --dump=keys` shows the expected line `XXX-YYY PASSWORD`. Readding the key with `ocat --load` also didn't help.

Does anyone have an idea what happened and how to fix this?

**Comment by jpmens** [comment]
@harbra sorry to hear you're experiencing problems, and we've never come across that one. The diagnostic `no decryption key for XXX-YYY` definitely means the key cannot be retrieved, and thus (as you've experienced) there'll be no decryption. Have you verified that your key name (`XXX-YYY`) doesn't contain any trailing space or unprintable characters? 

When you say you changed the MQTT "server domain", you mean its hostname or address? That has no influence on the decryption phase.

Please treble-check that your device has exactly the same key as that which is configured on the ot-recorder side. Do you notice any difference? Pay particular attention to spaces in the names ...

**Comment by harbra** [comment]
@jpmens Thanks for the quick feedback and the hints. I quadruple-checked the output of `ocat --dump=keys` and there is no additional space, etc. I even changed the password in the Owntracks iOS app and ot-recorder to something easy (123) and added this to the LMDB database, to no avail.

Also changing the DeviceID and adding the "123" password for this user-deviceid combination did not work. It shows up in `ocat --dump=keys`, but the `no decryption key` message is still coming.

I'm at a loss here what's happening...:-( Do you have any other ideas how to further troubleshoot? Is there a way to reset the configuration and reinstalling ot-recorder without loosing all of my location history? Can I safely delete the files in the spool\owntracks\ghash dir?

**Comment by jpmens** [comment]
I recommend you proceed as follows: stop the recorder and create a backup of your whole SPOOL directory. Then go ahead and remove the files in `spool/owntracks/ghash/*` and initialize them again with `ot-recorder --initialize`. 

If you want to attempt to re-instate encryption, add your keys to the database; you already know how to load those with `ocat`. Launch `ot-recorder` and keep an eye on the tail of `store/owntracks/rec/<username>/<devicename>/2020-07.rec` while you hit publish on a device; you should see a JSON blob with your location data (unencrypted of course).

If that still doesn't work, disable the encryption setting on the device (which is this actually, iOS or Android?) and try again.

In doing this you will not lose the content of your location history (that is stored in files in `rec/*/*/*.rec`). However, entries which could not be decrypted still are.

**Comment by jpmens** [comment]
@harbra would you please also show us the output of `ocat --version`?

**Comment by harbra** [comment]
@jpmens I did as adviced (remove everything in `ghash/`, initialize again), however still no luck an same error message. Pushing from the iOS app unencrypted works as expected and gives me a JSON blob in `2020-07.rec`.

This is the output of `ocat --version`:
```
This is OwnTracks Recorder, version 0.8.6
built with:
        WITH_MQTT = yes
        WITH_HTTP = yes
        WITH_LUA  = yes
        WITH_ENCRYPT = yes
        WITH_PING = yes
        WITH_KILL = yes
        CONFIGFILE = "/etc/default/ot-recorder"
        STORAGEDEFAULT = "/var/spool/owntracks/recorder/store"
        STORAGEDIR = "/var/spool/owntracks/recorder/store"
        DOCROOT = "/usr/share/owntracks/recorder/htdocs"
        GHASHPREC = 7
        DEFAULT_HISTORY_HOURS = 6
        JSON_INDENT = "NULL"
        LIBMOSQUITTO_VERSION = 1.5.7
        MDB VERSION = LMDB 0.9.22: (March 21, 2018)
        SODIUM VERSION = 1.0.17
        GIT VERSION = 0.8.6-0-gd468cba305
```

**Comment by jpmens** [comment]
I'm going to start grasping at straws ... 

1. Which OS/version is ot-recorder running on?
2. How did you install it? (i.e. from a package or self-compiled)
3. You write above `XXX-YYY in owntracks/XXX/YYY`. Are `XXX` and `YYY` really in capital letters? 

We're also going to have to do this please in order for me to try to see what's going on:

1. please change the encryption password on iOS and via `ocat` to something you don't mind sharing with me
2. publish a position or two or three from iOS
3. subscribe to your MQTT broker with `mosquitto_sub` and grab whatever is shown on _stdout_
4. grab the logs on when ot-recorder starts and as it receives the three publishes from iOS
5. grab the tail end of the appropriate `.rec` file

and if you pack all that together (zip or tar is fine) and send me that at `support@owntracks.org` I will take a look.

**Comment by harbra** [comment]
First of all thanks for all the help with troubleshooting!

Regarding your questions:
1. I'm using Raspbian buster on a Raspberry Pi 3B.
2. ot-recorder was installed with apt-get from http://repo.owntracks.org as documented here on github.
3. No, I'm using all lowercase letters. The capital XXX-YYY was just used to illustrate that these are not my actual user and device.

I also did the steps you described and sent the logs to your support address. I already saw something in the logs that I would assume is causing this issue, however: `gcache_open: mdb_env_open: Cannot allocate memory`. I assume the keys can't be loaded from the database and then are not found when they are needed. However I'm not sure why this would happen. `top` shows >200MB free memory and `data.mdb` only has a size of 45kB.

**Comment by jpmens** [comment]
Thank you for the files, @harbra, and I apologize: I forgot to say I also need the `gcache/` directory in full. Note that you please have to publish a position from your phone with a password which matches what you will send me contained in `gcache/`

 You are right: `gcache_open: mdb_env_open: Cannot allocate memory` is not good at all, and I also see `Cannot read defaults from /etc/default/ot-recorder: Permission denied` which is not tragic, but you should please fix by adjusting the permissions on that file.

I have to ask again: when you removed everything in `ghash/` and initialized again with `ot-recorder --initialize` and you then launched the recorder, did you see the `gcache_open: mdb_env_open` error? Could you verify that if necessary?

**Comment by harbra** [comment]
Sorry for the late reply. I will send the gchache directory over after finishing this post.

I think the `Cannot read defaults from /etc/default/ot-recorder: Permission denied` can be neglected. I ran ot-recorder once as my local user and not the `owntracks` user. The local once doesn't have access to the owntracks files.

Regarding `gcache_open: mdb_env_open: Cannot allocate memory`: I just cleared the `ghash` directory again and re-initialized and the same error shows up in the logs, even before adding the decryption key.

**Comment by jpmens** [comment]
Thank you for the files which, unfortunately, don't tell me much.

I think we're going to have to go a bit deeper, but I need your help as I don't have the hardware. 

Can you install ot-recorder from source and try with that? It's not super difficult, and I think (hope!) we have that documented properly in the README. 

I'd like you reproduce the error you're seeing.

**Comment by harbra** [comment]
Dear jpmens. I installed ot-recorder from source (the documentation was almost spot on;-)), however the same problem occured.

I then dug a little in the source code and modified `#define LMDB_DB_SIZE	((size_t)100000 * (size_t)(1024 * 1024))` in `gcache.h`. If I interpret the line correctly, it tries to allocate 100GB for the database. I changed this to ``#define LMDB_DB_SIZE	((size_t)50 * (size_t)(1024 * 1024))`, rebuilt and it works!

However I'm not sure how this can happen as apparently the `LMDB_DB_SIZE` has been at the given value since forever and as I wrote initially, `ot-recorder` worked without isses until around April...

**Comment by jpmens** [comment]
That was quick, @harbra; well done, and you did exactly what I was going to ask you to do. That size is just maximum size of the database, it's not a region that is actually allocated, but I did fear that that was the issue.

However, as you already said, you've been running with that since forever until April, so we know it cannot be that... (your file system isn't full, I take it)

I've really no idea other than asking you whether in or around April you did system upgrades, etc. That's all which occurs to me.

If you've no issue with running the self-compiled version that would, at least for the time being, permit you to go productive.

**Comment by harbra** [comment]
I tracked down the exact date that encryption stated failing. I really did a system upgrade on that day moving to Raspian Buster. I noticed in the apt logs that also ot-recorder was updated from 0.7.9 to 0.8.6. It seems a fix that was included in 0.8.0 (#253) indeed did increase `LMDB_DB_SIZE` from 120MB to 100GB (https://github.com/owntracks/recorder/compare/0.7.9...0.8.6#diff-0df380d6b1c36349bfe70f1483f58974). Maybe it is a peculiarity of the Raspbian operating system that an excessive maximum DB size causes issues? But then I would have expected others to run into the same issue...

Anyways, thanks for all your help, I'll then just stick to the self compiled version.

**Comment by nikitaindik** [comment]
Hi @jpmens! I am facing the same issue when running the Recorder on a Raspberry Pi in Docker using this Dockerfile: https://github.com/owntracks/docker-recorder/blob/master/Dockerfile

Changing `LMDB_DB_SIZE` to a more modest value and recompiling does the trick, but it would be nice to be able to pass this value through config file or using env variables. What do you think?

**Comment by jpmens** [comment]
I agree and have re-opened this issue.

**Comment by pablito1755** [comment]
I ran into this as well, but was happy I landed here. As I wanted to get recorder running as a service and had already built 0.9.0 from source, I first tried adding _OTR_LMDBSIZE=125829120_ to /etc/default/ot-recorder, but that didn't have the desired effect. I then tried adding _Environment="OTR_LMDBSIZE=125829120"_ to /etc/systemd/system/ot-recorder.service and this did the trick thanks to the recent fix added by jpmens.

Thanks!
